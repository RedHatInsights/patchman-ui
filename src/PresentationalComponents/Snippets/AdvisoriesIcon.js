import React from 'react';
import { Flex, FlexItem, Tooltip } from '@patternfly/react-core';
import propTypes from 'prop-types';

const AdvisoriesIcon = ({ count, tooltipText, Icon }) =>(
    <Tooltip content={tooltipText}>
        <Flex flex={{ default: 'inlineFlex' }}>
            <FlexItem spacer={{ default: 'spacerSm' }}>
                <Icon/>
            </FlexItem>
            <FlexItem spacer={{ default: 'spacerSm' }}>
                {count.toString()}
            </FlexItem>
        </Flex>
    </Tooltip>
);

AdvisoriesIcon.propTypes = {
    Icon: propTypes.element,
    count: propTypes.any,
    tooltipText: propTypes.string
};

export default AdvisoriesIcon;
