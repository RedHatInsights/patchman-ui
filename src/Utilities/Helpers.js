import { Tooltip } from '@patternfly/react-core';
import BugIcon from '@patternfly/react-icons/dist/js/icons/bug-icon';
import EnhancementIcon from '@patternfly/react-icons/dist/js/icons/enhancement-icon';
import SecurityIcon from '@patternfly/react-icons/dist/js/icons/security-icon';
import { SortByDirection } from '@patternfly/react-table/dist/js';
import findIndex from 'lodash/findIndex';
import qs from 'query-string';
import React from 'react';
import { Link } from 'react-router-dom';
import { advisorySeverities, filterCategories } from './constants';

export const convertLimitOffset = (limit, offset) => {
    return [offset / limit + 1, limit];
};

export const createSortBy = (header, values, offset) => {
    if (values) {
        let [value] = values;
        let direction =
            value[0] === '-' ? SortByDirection.desc : SortByDirection.asc;
        value = value.replace(/^(-|\+)/, '');
        const index = findIndex(header, item => item.key === value);
        let sort = {
            index: index + offset,
            direction
        };
        return sort;
    }

    return {};
};

export const addOrRemoveItemFromSet = (targetObj, inputArr) => {
    const inputObj = inputArr.reduce(
        (obj, item) => ((obj[item.rowId] = item.value || undefined), obj),
        {}
    );
    const result = { ...targetObj, ...inputObj };
    return result;
};

export const getNewSelectedItems = (selectedItems, currentItems) => {
    let payload = [].concat(selectedItems).map(item=>({ rowId: item.id, value: item.selected }));
    return addOrRemoveItemFromSet(
        currentItems,
        payload
    );
};

// for expandable rows only
export const getRowIdByIndexExpandable = (arrayOfObjects, index) => {
    return arrayOfObjects[index / 2].id;
};

export const getOffsetFromPageLimit = (page, limit) => {
    return page * limit - limit;
};

export const getLimitFromPageSize = limit => {
    return limit;
};

export function truncate(str, max, end) {
    return str.length > max ? (
        <React.Fragment>
            {str.substr(0, max - 1)}
            ...&nbsp;{end}
        </React.Fragment>
    ) : str;
}

export function createAdvisoriesIcons([rhea, rhba, rhsa]) {
    return (
        <div>
            {[rhea, rhba, rhsa].every(item => item === 0) &&
                'No applicable advisories'}
            {rhea !== 0 && (
                <Tooltip content={'Enhancements'}>
                    <span className="icon-with-label">
                        <EnhancementIcon />
                        {rhea.toString()}
                    </span>
                </Tooltip>
            )}
            {rhba !== 0 && (
                <Tooltip content={'Bug fixes'}>
                    <span className="icon-with-label">
                        <BugIcon />
                        {rhba.toString()}
                    </span>
                </Tooltip>
            )}
            {rhsa !== 0 && (
                <Tooltip content={'Security advisories'}>
                    <span className="icon-with-label">
                        <SecurityIcon
                            color={'var(--pf-global--warning-color--100)'}
                        />
                        {rhsa.toString()}
                    </span>
                </Tooltip>
            )}
        </div>
    );
}

export function getSeverityById(id) {
    return (
        advisorySeverities.find(item => item.value === id) ||
        advisorySeverities[0]
    );
}

export function handleAdvisoryLink(name, body) {
    if (location.href.indexOf('inventory') === -1) {
        return (
            <Link to={'/advisories/' + name}>
                {body === undefined ? name : body}
            </Link>
        );
    } else {
        return (
            <a href={`${document.baseURI}insights/patch/advisories/${name}`}>
                {body || name}
            </a>
        );
    }
}

export const arrayFromObj = items =>
    Object.keys(items).filter(key => items[key]);

export const remediationProvider = (issues, systems) => {
    issues = [].concat(issues);
    systems = [].concat(systems);
    return issues.length && systems.length
        ? {
            issues: issues.map(item => ({
                id: `patch-advisory:${item}`,
                description: item
            })),
            systems
        }
        : false;
};

export const getFilterValue = (category, key) => {
    const filterCategory = filterCategories[category];
    if (filterCategory) {
        const filterOption = filterCategory.values.find(
            item => item.value === key
        );
        return filterOption || { apiValue: key };
    }
};

export const encodeParams = (parameters, shouldTranslateKeys) => {
    const flattenFilters = filter => {
        let result = {};
        filter &&
            Object.entries(filter).forEach(item => {
                let [key, value] = item;
                value = shouldTranslateKeys && getFilterValue(key, value).apiValue || value;
                const operator = [].concat(value).length > 1 ? 'in:' : '';
                result = {
                    ...result,
                    [`filter[${key}]`]: `${operator}${value.toString()}`
                };
            });
        return result;
    };

    let { filter, ...allParams } = parameters;
    allParams = { ...allParams, ...flattenFilters(filter) };
    let params = [];
    Object.keys(allParams).forEach(key => {
        const argKey = encodeURIComponent(key);
        const argValue = encodeURIComponent(allParams[key]);
        if (argValue !== '') {
            params.push(argKey.concat('=').concat(argValue));
        }
    });

    return '?'.concat(params.join('&'));
};

export const encodeApiParams = parameters => {
    return encodeParams(parameters, true);
};

export const encodeURLParams = parameters => {
    delete parameters.id;
    return encodeParams(parameters, false);
};

export const decomposeFilterValue = filterValue => {
    if (filterValue.startsWith('in:')) {
        const values = filterValue.slice(3);
        return values.split(',');
    }

    return filterValue;
};

export const decodeQueryparams = queryString => {
    const parsed = qs.parse(queryString);
    const res = {};
    Object.keys(parsed).forEach(key => {
        const value = parsed[key];
        const bracketIndex = key.search(/\[.*\]/);
        if (bracketIndex > 0) {
            const objParent = key.slice(0, bracketIndex);
            const objKey = key.slice(bracketIndex + 1, -1);
            res[objParent] = {
                ...res[objParent],
                [objKey]: decomposeFilterValue(value)
            };
        } else {
            res[key] = value;
        }
    });
    return res;
};

export const buildFilterChips = (filters, search) => {
    let filterConfig = [];
    const processFilters = () => {
        let categories = Object.keys(filters).filter(
            item =>
                filters[item] !== '' && [].concat(filters[item]).length !== 0
        );
        filterConfig = filterConfig.concat(
            categories.map(category => {
                const { label, values } = filterCategories[category];
                return {
                    category: label,
                    id: category,
                    chips: [].concat(filters[category]).map(filterValue => {
                        const match = values.find(
                            item =>
                                item.value.toString() === filterValue.toString()
                        );
                        return {
                            name: match.label,
                            value: filterValue,
                            id: match.value
                        };
                    })
                };
            })
        );
    };

    const processSearch = () => {
        filterConfig = filterConfig.concat([
            {
                category: 'Search',
                id: 'search',
                chips: [
                    {
                        name: search,
                        value: search
                    }
                ]
            }
        ]);
    };

    filters && processFilters();
    search && processSearch();

    return filterConfig;
};

export const changeListParams = (oldParams, newParams) => {
    const newState = { ...oldParams, ...newParams };
    const offsetResetParams = ['filter', 'search', 'limit'];
    if (offsetResetParams.some(item => newParams.hasOwnProperty(item))) {
        newState.offset = 0;
    }

    if (newParams.hasOwnProperty('filter')) {
        newState.filter = { ...oldParams.filter, ...newParams.filter };
    }

    return newState;
};

export function subtractDate(days) {
    const date = new Date();
    date.setDate(date.getDate() - days);
    return date.toISOString();
}
