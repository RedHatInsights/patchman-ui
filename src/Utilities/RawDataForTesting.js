// raw data for testing purposes only

export const advisoryRows = [{
    attributes: {
        advisory_type: 1,
        applicable_systems: 1,
        description: 'The microcode_ctl ',
        public_date: '2020-06-24T17:22:25Z',
        synopsis: 'microcode_ctl',
        severity: 3
    },
    id: 'RHEA-2020:2743',
    type: 'advisory'
}];

export const systemAdvisoryRows = [{
    attributes: {
        advisory_type: 3,
        description: 'Kernel-based Virtual.',
        public_date: '2020-06-30T13:38:53Z',
        severity: 3,
        synopsis: 'Important'
    },
    id: 'RHSA-2020:2774',
    type: 'advisory'
}];

export const systemRows = [{
    attributes: {
        display_name: 'automation_host',
        last_evaluation: '2020-08-12T13:57:54.028883Z',
        last_upload: '2020-08-12T09:26:33.891907Z',
        packages_installed: 0,
        packages_updatable: 0,
        rhba_count: 213,
        rhea_count: 35,
        rhsa_count: 77,
        stale: false
    },
    id: '8ddb54f5-aeeb-49b1-8448-c29049d686c1',
    type: 'system'
}];

export const advisoryHeader = {
    attributes: {
        advisory_type: 3,
        description: 'Kernel-based Virtual.',
        public_date: '2020-06-30T13:38:53Z',
        modified_date: '2020-06-30T13:38:53Z',
        severity: 3,
        synopsis: 'Important'
    },
    id: 'RHSA-2020:2774',
    type: 'advisory'
};

export const advisoryDetailRows = {
    data: {
        attributes: {
            cves: ['CVE-2020-11080'],
            description: 'libnghttp2.',
            fixes: null,
            modified_date: '2020-06-25T16:31:29Z',
            packages: { libnghttp2: '1.33.0-3.el8_2.1.x86_64' },
            public_date: '2020-06-25T16:31:29Z',
            references: [],
            severity: 3,
            solution: 'For details',
            synopsis: 'Important: ',
            topic: 'An update .'
        }
    },
    id: 'RHSA-2020:2755',
    type: 'advisory'
};
export const systemPackages = [{
    description: 'This package contains testing data',
    evra: '2.2.*',
    name: 'acl',
    summary: 'Access control list utilities',
    updates: null
}];

export const entityDetail = {
    loaded: true,
    activeApps: [],
    entity: {
        account: '60'
    }
};
