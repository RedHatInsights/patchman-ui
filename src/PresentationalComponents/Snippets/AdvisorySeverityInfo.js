import { Text, TextContent, TextVariants, Popover } from '@patternfly/react-core';
import { ExternalLinkSquareAltIcon } from '@patternfly/react-icons';
import propTypes from 'prop-types';
import React from 'react';
import { Flex, FlexItem } from '@patternfly/react-core';
import { intl } from '../../Utilities/IntlProvider';
import messages from '../../Messages';

const AdvisorySeverityInfo = ({ severity }) => {
    return (
        <Popover
            position="bottom"
            enableFlip
            headerContent={<div>{severity.label + ' severity'}</div>}
            bodyContent={
                <TextContent>
                    <Text component={TextVariants.p}>{severity.text}</Text>
                </TextContent>
            }
            footerContent={
                <a
                    href="https://access.redhat.com/security/updates/classification/"
                    target="_blank"
                    rel="noopener noreferrer"
                >
                    <Flex>
                        <FlexItem spacer={{ default: 'spacerSm' }}>
                            <ExternalLinkSquareAltIcon />
                        </FlexItem>
                        <FlexItem spacer={{ default: 'spacerSm' }}>
                            {intl.formatMessage(messages.linksSearchSecurityRatings)}
                        </FlexItem>
                    </Flex>
                </a>
            }
        >
            <a>{intl.formatMessage(messages.linksLearnMore)}</a>
        </Popover>
    );
};

AdvisorySeverityInfo.propTypes = {
    severity: propTypes.object
};

export default AdvisorySeverityInfo;
