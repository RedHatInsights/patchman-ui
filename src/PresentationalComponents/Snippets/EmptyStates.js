import {
    EmptyState,
    EmptyStateIcon,
    EmptyStateVariant,
    Title
} from '@patternfly/react-core';
import SearchIcon from '@patternfly/react-icons/dist/js/icons/search-icon';
import React from 'react';
import { intl } from '../../Utilities/IntlProvider';
import messages from '../../Messages';

export const EmptyAdvisoryList = () => (
    <EmptyState variant={EmptyStateVariant.full}>
        <EmptyStateIcon icon={SearchIcon} />
        <Title headingLevel="h5" size="lg">
            {intl.formatMessage(messages.statesNoMatchingAdvisories)}
        </Title>
    </EmptyState>
);

export const EmptyPackagesList = () => (
    <EmptyState variant={EmptyStateVariant.full}>
        <EmptyStateIcon icon={SearchIcon} />
        <Title headingLevel="h5" size="lg">
            {intl.formatMessage(messages.statesNoMatchingPackages)}
        </Title>
    </EmptyState>
);

export const EmptyCvesList = () => (
    <EmptyState variant={EmptyStateVariant.full}>
        <EmptyStateIcon icon={SearchIcon} />
        <Title headingLevel="h5" size="lg">
            {intl.formatMessage(messages.statesNoMatchingCve)}
        </Title>
    </EmptyState>
);
