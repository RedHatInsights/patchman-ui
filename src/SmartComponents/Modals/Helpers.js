import React from 'react';
import { GridItem } from '@patternfly/react-core';

import messages from '../../Messages';
import { fetchSystems } from '../../Utilities/api';

export const filterSystemsWithoutSets = (systemsIDs) =>  {
    return fetchSystems({ limit: -1, 'filter[baseline_name]': 'neq:' }).then((allSystemsWithPatchSet) => {
        return systemsIDs.filter(systemID =>
            allSystemsWithPatchSet?.data.some(system => system.id === systemID)
        );
    });
};

export const renderUnassignModalMessages = (bodyMessage, systemsCount, intl) => (<GridItem>
    {intl.formatMessage(
        messages[bodyMessage],
        { systemsCount, b: (...chunks) => <b>{chunks}</b> }
    )}
</GridItem>);

