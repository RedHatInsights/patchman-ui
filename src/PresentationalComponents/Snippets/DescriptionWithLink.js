import React from 'react';
import { Text, TextContent, TextVariants, TextList,
    TextListVariants, TextListItem, TextListItemVariants } from '@patternfly/react-core';
import Label from './Label';
import { handlePatchLink, truncate, getSeverityById } from '../../Utilities/Helpers';
import { entityTypes } from '../../Utilities/constants';
import ExternalLink from './ExternalLink';
import propTypes from 'prop-types';
import { intl } from '../../Utilities/IntlProvider';
import messages from '../../Messages';
import { SecurityIcon } from '@patternfly/react-icons';

export const DescriptionWithLink = ({ row }) =>  {
    const severityObject = getSeverityById(row.attributes.severity);
    return (
        <TextContent>
            {
                row.attributes.cve_count > 0 &&
                (<TextList component={TextListVariants.dl} style ={{ '--pf-c-content--dl--RowGap': '0.5rem' }}>
                    <TextListItem component={TextListItemVariants.dt}>
                        {intl.formatMessage(messages.labelsSeverity)}
                    </TextListItem>
                    <TextListItem component={TextListItemVariants.dd}>
                        <SecurityIcon size="sm" color = {severityObject.color}/>  {severityObject.label}
                    </TextListItem>
                    <TextListItem component={TextListItemVariants.dt}>
                        {intl.formatMessage(messages.labelsCves)}
                    </TextListItem>
                    <TextListItem component={TextListItemVariants.dd}>
                        {row.attributes.cve_count}
                    </TextListItem>
                </TextList>)}
            <Label>{intl.formatMessage(messages.labelsDescription)}</Label>
            <Text component={TextVariants.p} style={{ whiteSpace: 'pre-line' }}>
                {truncate(row.attributes.description.replace(
                    new RegExp('\\n(?=[^\\n])', 'g'),
                    ''
                ), 570, handlePatchLink(entityTypes.advisories, row.id, intl.formatMessage(messages.linksReadMore)))}
            </Text>
            <ExternalLink link={`https://access.redhat.com/errata/${row.id}`}
                text={intl.formatMessage(messages.linksViewPackagesAndErrata)} />
        </TextContent>);
};

DescriptionWithLink.propTypes = {
    row: propTypes.shape({
        id: propTypes.string,
        attributes: propTypes.object
    })
};
