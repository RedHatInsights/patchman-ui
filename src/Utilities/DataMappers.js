import { processDate } from '@redhat-cloud-services/frontend-components-utilities/files/cjs/helpers';
import { flatMap } from 'lodash';
import React from 'react';
import AdvisoryType from '../PresentationalComponents/AdvisoryType/AdvisoryType';
import { EmptyAdvisoryList, EmptyPackagesList } from '../PresentationalComponents/Snippets/EmptyStates';
import { entityTypes } from './constants';
import { createUpgradableColumn, handlePatchLink } from './Helpers';
import { DescriptionWithLink } from '../PresentationalComponents/Snippets/DescriptionWithLink';

export const createAdvisoriesRows = (rows, expandedRows, selectedRows) => {
    if (rows.length !== 0) {
        return flatMap(rows, (row, index) => {
            return [
                {
                    id: row.id,
                    isOpen: expandedRows[row.id] === true,
                    selected: selectedRows[row.id] !== undefined,
                    cells: [
                        { title: handlePatchLink(entityTypes.advisories, row.id) },
                        { title: processDate(row.attributes.public_date) },
                        {
                            title: (
                                <AdvisoryType
                                    type={row.attributes.advisory_type}
                                />
                            )
                        },
                        {
                            title: handlePatchLink(
                                entityTypes.advisories,
                                row.id,
                                row.attributes.applicable_systems
                            )
                        },
                        row.attributes.synopsis
                    ]
                },
                {
                    cells: [
                        {
                            title: <DescriptionWithLink row = {row} />
                        }
                    ],
                    parent: index * 2
                }
            ];
        });
    } else {
        return [
            {
                heightAuto: true,
                cells: [
                    {
                        props: { colSpan: 5 },
                        title: <EmptyAdvisoryList />
                    }
                ]
            }
        ];
    }
};

export const createSystemAdvisoriesRows = (
    rows,
    expandedRows,
    selectedRows
) => {
    if (rows.length !== 0) {
        return flatMap(rows, (row, index) => {
            return [
                {
                    id: row.id,
                    isOpen: expandedRows[row.id] === true,
                    selected: selectedRows[row.id] !== undefined,
                    cells: [
                        { title: handlePatchLink(entityTypes.advisories, row.id) },
                        { title: processDate(row.attributes.public_date) },
                        {
                            title: (
                                <AdvisoryType
                                    type={row.attributes.advisory_type}
                                />
                            )
                        },
                        row.attributes.synopsis
                    ]
                },
                {
                    cells: [
                        {
                            title: <DescriptionWithLink row = {row} />
                        }
                    ],
                    parent: index * 2
                }
            ];
        });
    } else {
        return [
            {
                heightAuto: true,
                cells: [
                    {
                        props: { colSpan: 6 },
                        title: <EmptyAdvisoryList />
                    }
                ]
            }
        ];
    }
};

export const createSystemsRows = (rows, selectedRows = {}) => {
    const data =
        rows &&
        rows.map(row => {
            return {
                id: row.id,
                key: Math.random().toString() + row.id,
                packages_installed: row.attributes.packages_installed,
                applicable_advisories: [
                    row.attributes.rhea_count || 0,
                    row.attributes.rhba_count || 0,
                    row.attributes.rhsa_count || 0
                ],
                selected: selectedRows[row.id] !== undefined
            };
        });
    return data || [];
};

export const createPackageSystemsRows = (rows, selectedRows = {}) => {
    const data =
        rows &&
        rows.map(row => {
            return {
                id: row.id,
                key: Math.random().toString() + row.id,
                installed_evra: row.installed_evra,
                available_evra: row.updatable && row.available_evra || row.installed_evra,
                upgradable: row.updatable,
                selected: selectedRows[row.id] !== undefined
            };
        });
    return data || [];
};

export const createSystemPackagesRows = (rows, selectedRows = {}) => {
    if (rows.length !== 0) {
        return rows.map(pkg => {
            const pkgUpdates = pkg.updates || [];
            const latestUpdate = pkgUpdates[pkgUpdates.length - 1];

            return {
                id: pkg.name,
                key: pkg.name,
                selected: selectedRows[pkg.name] !== undefined,
                disableCheckbox: !latestUpdate,
                cells: [
                    { title: handlePatchLink(entityTypes.packages, pkg.name) },
                    { title: pkg.evra },
                    { title: (latestUpdate && latestUpdate.evra) || pkg.evra },
                    { title: createUpgradableColumn(pkg.updatable) },
                    { title: pkg.summary }
                ]
            };
        });
    } else {
        return [
            {
                heightAuto: true,
                cells: [
                    {
                        props: { colSpan: 7 },
                        title: <EmptyPackagesList />
                    }
                ]
            }
        ];
    }
};

export const createPackagesRows = (rows) => {
    if (rows.length !== 0) {
        return rows.map(pkg => {
            return {
                id: pkg.name,
                key: pkg.name,
                cells: [
                    { title: handlePatchLink(entityTypes.packages, pkg.name) },
                    { title: pkg.systems_installed },
                    { title: pkg.systems_updatable },
                    { title: pkg.summary }
                ]
            };
        });
    } else {
        return [
            {
                heightAuto: true,
                cells: [
                    {
                        props: { colSpan: 7 },
                        title: <EmptyPackagesList />
                    }
                ]
            }
        ];
    }
};
