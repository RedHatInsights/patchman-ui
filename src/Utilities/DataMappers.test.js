import { processDate } from '@redhat-cloud-services/frontend-components-utilities/files/cjs/helpers';
import { EmptyAdvisoryList } from '../PresentationalComponents/Snippets/EmptyStates';
import { createAdvisoriesRows, createSystemAdvisoriesRows, createSystemPackagesRows, createSystemsRows } from './DataMappers';
import { handlePatchLink } from './Helpers';
import { advisoryRows, systemAdvisoryRows, systemPackages, systemRows } from './RawDataForTesting';

describe('DataMappers', () => {
    it('Should create advisories rows', () => {
        const [firstRow, secondRow] = createAdvisoriesRows(advisoryRows, [], []);
        expect(firstRow.id).toEqual(advisoryRows[0].id);
        expect(firstRow.isOpen).toEqual(false);
        expect(firstRow.selected).toEqual(false);
        expect(firstRow.cells[0].title).toEqual(handlePatchLink('advisories', advisoryRows[0].id));
        expect(firstRow.cells[1].title).toEqual(processDate(advisoryRows[0].attributes.public_date));
        expect(firstRow.cells[2].title.props.type).toEqual(advisoryRows[0].attributes.applicable_systems);
        expect(firstRow.cells[3].title).toEqual(
            handlePatchLink('advisories', advisoryRows[0].id, advisoryRows[0].attributes.applicable_systems)
        );
        expect(firstRow.cells[4]).toEqual(advisoryRows[0].attributes.synopsis);
        const portalAdvisoryLink = secondRow.cells[0].title;
        expect(portalAdvisoryLink.props.row).toEqual(advisoryRows[0]);
    });

    it('Should createAdvisoriesRows handle empty row data', () => {
        const result = createAdvisoriesRows([], [], []);
        expect(result[0].heightAuto).toBeTruthy();
        expect(result[0].cells[0].props).toEqual({ colSpan: 5 });
        expect(result[0].cells[0].title.type).toEqual(EmptyAdvisoryList);
    });

    it('Should create System Advisories Rows', () => {
        const [firstRow, secondRow] = createSystemAdvisoriesRows(systemAdvisoryRows, [], []);
        expect(firstRow.id).toEqual(systemAdvisoryRows[0].id);
        expect(firstRow.isOpen).toEqual(false);
        expect(firstRow.selected).toEqual(false);
        expect(firstRow.cells[0].title).toEqual(handlePatchLink('advisories', systemAdvisoryRows[0].id));
        expect(firstRow.cells[1].title).toEqual(processDate(systemAdvisoryRows[0].attributes.public_date));
        expect(firstRow.cells[2].title.props.type).toEqual(systemAdvisoryRows[0].attributes.advisory_type);
        expect(firstRow.cells[3]).toEqual(systemAdvisoryRows[0].attributes.synopsis);
        const portalAdvisoryLink = secondRow.cells[0].title;
        expect(portalAdvisoryLink.props.row).toEqual(systemAdvisoryRows[0]);
    });

    it('Should createSystemAdvisoriesRows handle empty row data', () => {
        const result = createSystemAdvisoriesRows([], [], []);
        expect(result[0].heightAuto).toBeTruthy();
        expect(result[0].cells[0].props).toEqual({ colSpan: 6 });
        expect(result[0].cells[0].title.type).toEqual(EmptyAdvisoryList);
    });

    it('Should create system rows', () => {
        const result = createSystemsRows(systemRows, []);
        expect(result[0].id).toEqual(systemRows[0].id);
        expect(result[0].key).toEqual(expect.stringContaining(systemRows[0].id));
        expect(result[0].applicable_advisories).toEqual([
            systemRows[0].attributes.rhea_count,
            systemRows[0].attributes.rhba_count,
            systemRows[0].attributes.rhsa_count
        ]);
        expect(result[0].selected).toEqual(false);
    });

    it('Should createSystemRows handle empty data', () => {
        const result = createSystemsRows(undefined, []);
        expect(result).toEqual([]);
    });

    it('Should createSystemRows handle undfined rhea_count, rhba_count, rhsa_count', () => {
        const testData = [{
            attributes: {
                display_name: 'automation_host',
                last_evaluation: '2020-08-12T13:57:54.028883Z',
                last_upload: '2020-08-12T09:26:33.891907Z',
                packages_installed: 0,
                packages_updatable: 0,
                stale: false
            },
            id: '8ddb54f5-aeeb-49b1-8448-c29049d686c1',
            type: 'system'
        }];

        const result = createSystemsRows(testData);
        expect(result[0].applicable_advisories).toEqual([
            0,
            0,
            0
        ]);
    });

    it('Should create package list', () => {
        const packages = createSystemPackagesRows(systemPackages);
        expect(packages).toEqual([{
            id: 'acl',
            key: 'acl',
            selected: false,
            disableCheckbox: true,
            cells: [
                { title: expect.anything() }, // FIXME!
                { title: '2.2.*' },
                { title: '2.2.*' },
                { title: expect.anything() },
                { title: 'Access control list utilities' }
            ]
        }]);

        const packagesWithCompleteData = createSystemPackagesRows([{
            ...systemPackages[0],
            updates: [{ evra: 'testEvra' }],
            updatable: true
        }]);

        expect(packagesWithCompleteData).toEqual([{
            id: 'acl',
            key: 'acl',
            selected: false,
            disableCheckbox: false,
            cells: [
                { title: expect.anything()  }, // FIXME!
                { title: '2.2.*' },
                { title: 'testEvra' },
                { title: expect.anything() },
                { title: 'Access control list utilities' }
            ]
        }]);
    });

    it('Should create empty package list', () => {
        const packages = createSystemPackagesRows([]);
        expect(packages).toEqual([{
            heightAuto: true,
            cells: [
                {
                    props: { colSpan: 7 },
                    title: expect.anything()
                }
            ]
        }]);
    });
});
