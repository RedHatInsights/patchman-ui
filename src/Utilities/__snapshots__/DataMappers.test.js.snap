// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataMappers Should create package list 1`] = `
Object {
  "cells": Array [
    Object {
      "title": <Link
        to="/packages/test-name-2"
      >
        test-name-2
      </Link>,
    },
    Object {
      "title": "test-evra-2",
    },
    Object {
      "title": "test-update-evra",
    },
    Object {
      "title": <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
          }
        }
      >
        <CheckIcon
          color="currentColor"
          noVerticalAlign={false}
          size="sm"
          style={
            Object {
              "color": "var(--pf-global--success-color--100)",
            }
          }
        />
        <span
          style={
            Object {
              "marginLeft": "var(--pf-global--spacer--sm)",
            }
          }
        >
          Up-to-date
        </span>
      </div>,
    },
    Object {
      "title": "Access control list utilities",
    },
  ],
  "disableSelection": false,
  "id": "test-name-2-test-evra-2",
  "key": "test-name-2-test-evra-2",
  "selected": false,
}
`;
